xdem/coreg.py

```diff
@@ -1063,3 +1081,620 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
+    _fit_called = False  # Flag to check if the .fit() method has been called.
+
+    def __init__(self):
+        """This function should have been overwritten by subclassing."""
+        raise ValueError("Coreg class should not be instantiated directly.")
+
+    def fit(self, reference_dem: Union[np.ndarray, np.ma.masked_array],
+            dem_to_be_aligned: Union[np.ndarray, np.ma.masked_array],
+            inlier_mask: Optional[np.ndarray] = None,
+            transform: Optional[rio.transform.Affine] = None,
+            weights: Optional[np.ndarray] = None,
+            subsample: Union[float, int] = 1.0):
+        """
+        Estimate the coregistration transform on the given DEMs.
+
+        :param reference_dem: 2D array of elevation values acting reference. 
+        :param dem_to_be_aligned: 2D array of elevation values to be aligned.
+        :param inlier_mask: Optional. 2D boolean array of areas to include in the analysis (inliers=True).
+        :param transform: Optional. Transform of the reference_dem. Mandatory in some cases.
+        :param weights: Optional. Per-pixel weights for the coregistration.
+        :param subsample: Subsample the input to increase performance. <1 is parsed as a fraction. >1 is a pixel count.
+        """
+        # Make sure that the mask has an expected format.
+        if inlier_mask is not None:
+            inlier_mask = np.asarray(inlier_mask)
+            assert inlier_mask.dtype == bool, f"Invalid mask dtype: '{inlier_mask.dtype}'. Expected 'bool'"
+
+        if weights is not None:
+            raise NotImplementedError("Weights have not yet been implemented")
+
+        # The reference mask is the union of the nan occurrence and the (potential) ma mask.
+        ref_mask = np.isnan(reference_dem) | (reference_dem.mask
+                                              if isinstance(reference_dem, np.ma.masked_array) else False)
+        # The to-be-aligned mask is the union of the nan occurrence and the (potential) ma mask.
+        tba_mask = np.isnan(dem_to_be_aligned) | (dem_to_be_aligned.mask
+                                                  if isinstance(dem_to_be_aligned, np.ma.masked_array) else False)
+
+        # The full mask (inliers=True) is the inverse of the above masks and the provided mask.
+        full_mask = (~ref_mask & ~tba_mask & (np.asarray(inlier_mask) if inlier_mask is not None else True)).squeeze()
+
+        # If subsample is not equal to one, subsampling should be performed.
+        if subsample != 1.0:
+            # If subsample is less than one, it is parsed as a fraction (e.g. 0.8 => retain 80% of the values)
+            if subsample < 1.0:
+                subsample = int(np.count_nonzero(full_mask) * (1 - subsample))
+
+            # Randomly pick N inliers in the full_mask where N=subsample
+            random_falses = np.random.choice(np.argwhere(full_mask.flatten()).squeeze(), int(subsample), replace=False)
```

Comment:
It's great to think about a subsampling functionality. However, I see that the amount of data that is loaded is still the same. I would favor an option where only a fraction of the data was loaded instead, to reduce memory usage, but maybe that requires too much re-working?

Reply:
Mind if we add this as an issue for future improvement?

Reply:
Sure!